File: src/dynatraceOtelLogHook.ts
import {
  BeforeHookContext,
  EvaluationDetails,
  FlagValue,
  HookContext,
} from "@openfeature/core"
import { Hook } from "@openfeature/server-sdk"
import { Span, SpanKind, Tracer } from "@opentelemetry/api"
import { appMetadata } from "./otelSetup" // Import app metadata

class DynatraceOtelLogHook implements Hook {
  private name: string
  private tracer: Tracer
  private spans: WeakMap<HookContext, Span> = new WeakMap()

  constructor(tracer: Tracer) {
    this.name = "DynatraceOtelLogHook"
    this.tracer = tracer
  }

  before(hookContext: BeforeHookContext) {
    const span = this.tracer.startSpan(
      `feature_flag_evaluation.${hookContext.flagKey}`,
      {
        kind: SpanKind.SERVER,
      }
    )
    if (span) {
      span.setAttributes({
        "feature_flag.key": hookContext.flagKey,
        "feature_flag.value_type": hookContext.flagValueType,
        "feature_flag.flagset": hookContext.flagKey,
        "feature_flag.project": appMetadata.project,
        "feature_flag.environment": appMetadata._environment,
      })
      if (hookContext.clientMetadata?.name) {
        span.setAttributes({
          "openfeature.client.name": hookContext.clientMetadata.name,
        })
      }
      if (hookContext.providerMetadata?.name) {
        span.setAttributes({
          "openfeature.provider.name": "devcycle",
        })
      }
    }
    this.spans.set(hookContext, span)
  }

  finally(
    hookContext: HookContext,
    evaluationDetails: EvaluationDetails<FlagValue>
  ): void {
    const { value, variant, reason, errorCode, errorMessage } =
      evaluationDetails
    console.log("evaluationDetails", evaluationDetails)
    const span = this.spans.get(hookContext)

    if (span) {
      if (errorCode) {
        span.setAttributes({
          "feature_flag.error_code": errorCode,
        })
      }
      if (errorMessage) {
        span.setAttributes({
          "feature_flag.error_message": errorMessage,
        })
      }

      span.setAttributes({
        "feature_flag.value": String(value),
        "feature_flag.reason": reason || "",
      })
      if (variant) {
        span.setAttributes({
          "feature_flag.variant": variant,
        })
      }
      console.log("span", span)
      span.end()
    }
  }
}

export { DynatraceOtelLogHook }

File: java-backend/src/main/java/com/shopper/config/DynatraceOtelLogHook.java
package com.shopper.config;

import dev.openfeature.sdk.EvaluationContext;
import dev.openfeature.sdk.FlagEvaluationDetails;
import dev.openfeature.sdk.Hook;
import dev.openfeature.sdk.HookContext;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.SpanKind;
import io.opentelemetry.api.trace.Tracer;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

@RequiredArgsConstructor
@Slf4j
public class DynatraceOtelLogHook implements Hook<Object> {

    private final Tracer tracer;
    private final OpenTelemetryConfig.AppMetadata appMetadata;
    private final Map<HookContext<Object>, Span> spans = new ConcurrentHashMap<>();

    @Override
    public Optional<EvaluationContext> before(HookContext<Object> ctx, Map<String, Object> hints) {
        Span span = tracer.spanBuilder("feature_flag_evaluation." + ctx.getFlagKey())
                .setSpanKind(SpanKind.SERVER)
                .startSpan();

        if (span != null) {
            span.setAttribute("feature_flag.key", ctx.getFlagKey());
            span.setAttribute("feature_flag.value_type", ctx.getType().name());
            span.setAttribute("feature_flag.flagset", ctx.getFlagKey());
            span.setAttribute("feature_flag.project", appMetadata.getProject());
            span.setAttribute("feature_flag.environment", appMetadata.getEnvironmentId());

            if (ctx.getClientMetadata() != null && ctx.getClientMetadata().getName() != null) {
                span.setAttribute("openfeature.client.name", ctx.getClientMetadata().getName());
            }

            if (ctx.getProviderMetadata() != null && ctx.getProviderMetadata().getName() != null) {
                span.setAttribute("openfeature.provider.name", "devcycle");
            }
        }

        spans.put(ctx, span);
        return Optional.empty();
    }

    @Override
    public void after(HookContext<Object> ctx, FlagEvaluationDetails<Object> details, Map<String, Object> hints) {
        Span span = spans.remove(ctx);

        if (span != null) {
            log.debug("evaluationDetails: {}", details);

            if (details.getErrorCode() != null) {
                span.setAttribute("feature_flag.error_code", details.getErrorCode().toString());
            }

            if (details.getErrorMessage() != null) {
                span.setAttribute("feature_flag.error_message", details.getErrorMessage());
            }

            span.setAttribute("feature_flag.value", String.valueOf(details.getValue()));
            span.setAttribute("feature_flag.reason", details.getReason() != null ? details.getReason() : "");

            if (details.getVariant() != null) {
                span.setAttribute("feature_flag.variant", details.getVariant());
            }

            log.debug("Feature flag span completed: {}", ctx.getFlagKey());
            span.end();
        }
    }

    @Override
    public void error(HookContext<Object> ctx, Exception error, Map<String, Object> hints) {
        Span span = spans.remove(ctx);
        if (span != null) {
            span.recordException(error);
            span.setAttribute("feature_flag.error_message", error.getMessage());
            span.end();
        }
    }

    @Override
    public void finallyAfter(HookContext<Object> ctx, Map<String, Object> hints) {
        // Ensure span is always cleaned up
        Span span = spans.remove(ctx);
        if (span != null) {
            span.end();
        }
    }
}